File: Lab/src/Models/Member.ts
export interface Member {
  id?: number;
  cin: string;
  nom: string;
  prenom: string;
  date?: Date;
  photo?: string;
  cv?: string;
  email: string;
  password?: string;
  pubs?: Publication[];
  type?: string; // ← AJOUTER CE CHAMP
}

export interface Etudiant extends Member {
  type: 'etd';
  dateInscription?: Date;
  diplome: string;
  encadrant?: EnseignantChercheur;
}

export interface EnseignantChercheur extends Member {
  type: 'ens';
  grade: string;
  etablissement: string;
}
// Interface pour Publication
export interface Publication {
  id?: number;
  titre: string;
  type: string;
  dateApparition: Date;
  lien: string;
  sourcePdf?: string;
}

File: Projet Micro Service Backend/Member-Service/src/main/java/com/member/entities/Member.java
package com.member.entities;

import java.util.Collection;
import java.util.Date;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.member.beans.PublicationBean;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Inheritance;
import jakarta.persistence.InheritanceType;
import jakarta.persistence.DiscriminatorColumn;
import jakarta.persistence.DiscriminatorType;
import jakarta.persistence.Transient;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "type_mbr", discriminatorType = DiscriminatorType.STRING, length = 3)
@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.PROPERTY,
    property = "type",
    visible = true
)
@JsonSubTypes({
    @JsonSubTypes.Type(value = Etudiant.class, name = "etd"),
    @JsonSubTypes.Type(value = EnseignantChercheur.class, name = "ens")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public abstract class Member {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String cin;
    private String nom;
    private String prenom;
    private Date date;
    private String photo;
    private String cv;
    private String email;
    private String password;

    @Transient
    private Collection<PublicationBean> pubs;

    @Override
    public String toString() {
        return "Membre{" +
                "id=" + id +
                ", cin='" + cin + '\'' +
                ", nom='" + nom + '\'' +
                ", prenom='" + prenom + '\'' +
                ", date=" + date +
                ", cv='" + cv + '\'' +
                ", email='" + email + '\'' +
                '}';
    }
}